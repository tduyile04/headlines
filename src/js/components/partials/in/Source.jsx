import React from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import { withRouter } from 'react-router-dom';
import 'react-select/dist/react-select.css';
import HeadlineAction from '../../../actions/HeadlineAction';
import logo from '../../../../json/logo.json';

/**
 * Child Component detailing the display and attributes of each source component
 * @class Source
 * @extends {React.Component}
 */
class Source extends React.Component {
  constructor() {
    super();
    this.state = {
      logos: '',
      value: 'top'
    };
    this.showArticles = this.showArticles.bind(this);
  }

  componentWillMount() {
    this.setState({
      logos: logo
    });
  }

  /**
   * Shows a corresponding list of articles for a new source
   * @param {any} sourceName Name of news source viewed
   * @param {any} chosenSortOption Selected sort by options
   * @memberof Source
   */
  showArticles(sourceName, chosenSortOption) {
    HeadlineAction.saveSortOptions(chosenSortOption);
    this.props.history.push('/articles/' + sourceName + '/' + chosenSortOption);
  }

  /**
   * Allows change of the selected option to its new value
   * @param {any} newValue value of the selected sort option
   * @memberof Source
   */
  onChange(newValue) {
    this.setState({
      value: newValue.value
    });
  }

  render() {
    const { logos, value } = this.state;
    const source = this.props.source;
    const sortOptions = source.sortBysAvailable.map((option) => {
      return { value: option.toLowerCase(), label: option };
    });
    return (
        <div className="source col s12 m4">
          <div className="card card hoverable grey darken-3">
            <section className="row">
              <div className="col s4 m4">
                <img src={logos[source.id]} id="logo" />
              </div>
              <div className="col s8 m8">
                <h6 className="header white-text center-align">
                  {source.name}
                </h6>
              <div
              className="card-content"
              style = {{ padding: 0, height: 150, paddingLeft: 5 }}>
                <p
                className="grey-text text-lighten-5 content-description">
                  {source.description}
                </p>
              </div>
              </div>
            </section>
            <div className="card-stacked">
              <div className="card-action row">
                <div
                  className="orange-text text-accent-1 col m6 s6"
                  style={{ cursor: 'pointer' }}
                  onClick={
                    () => this.showArticles(
                      source.name, value
                    )
                  }
                >
                  View Articles
                </div>
                <div className="col m6 s6">
                <Select
                  name="sort-options"
                  value= {value}
                  options={sortOptions}
                  onChange={value => this.onChange(value)}
                />
                </div>
              </div>
            </div>
          </div>
        </div>
    );
  }
}

Source.propTypes = {
  source: PropTypes.object,
  history: PropTypes.object
};

export default withRouter(Source);
