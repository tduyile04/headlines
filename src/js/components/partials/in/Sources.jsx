import React from 'react';
import Source from './Source.jsx';

/**
 * Sources parent component for all the sources retrieved from the api
 * @class Sources
 * @extends {React.Component}
 */
class Sources extends React.Component {
  render() {
    const filteredSources = this.props.sources;
    return (
      <div className="row">
        {
          typeof filteredSources === 'object' &&
          filteredSources.map((source) => {
            return <Source key={source.id} source={source} />;
          }, this)
        }
      </div>
    );
  }
}

export default Sources;
