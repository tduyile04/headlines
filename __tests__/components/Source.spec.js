import React from 'react';
import { shallow } from 'enzyme';
import Source from '../../src/js/components/partials/in/Source.jsx';
import Sources from '../../src/js/components/partials/in/Sources.jsx';
import mockSources from '../../src/__mock__/sources.json';

describe('<Source />', () => {
  it('should display the same view for each render', () => {
    const wrapper = shallow(<Source />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render 1 <Sources />', () => {
    const sources = mockSources.sources;
    const wrapper = shallow(<Sources sources={sources}/>);
    expect(wrapper).toMatchSnapshot();
  });
});
