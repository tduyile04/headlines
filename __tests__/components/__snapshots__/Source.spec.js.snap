// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Source /> should display the same view for each render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Route
    render={[Function]}
/>,
  "nodes": Array [
    <Route
      render={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <withRouter(Source) />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Route
          render={[Function]}
/>,
        "_debugID": 2,
        "_renderedOutput": <Route
          render={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <withRouter(Source) />,
}
`;

exports[`<Source /> should render 1 <Sources /> 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="row"
>
    <withRouter(Source)
        source={
            Object {
                "category": "general",
                "country": "au",
                "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
                "id": "abc-news-au",
                "language": "en",
                "name": "ABC News (AU)",
                "sortBysAvailable": Array [
                  "top",
                ],
                "url": "http://www.abc.net.au/news",
                "urlsToLogos": Object {
                  "large": "",
                  "medium": "",
                  "small": "",
                },
              }
        }
    />
    <withRouter(Source)
        source={
            Object {
                "category": "general",
                "country": "us",
                "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
                "id": "al-jazeera-english",
                "language": "en",
                "name": "Al Jazeera English",
                "sortBysAvailable": Array [
                  "top",
                  "latest",
                ],
                "url": "http://www.aljazeera.com",
                "urlsToLogos": Object {
                  "large": "",
                  "medium": "",
                  "small": "",
                },
              }
        }
    />
    <withRouter(Source)
        source={
            Object {
                "category": "technology",
                "country": "us",
                "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
                "id": "ars-technica",
                "language": "en",
                "name": "Ars Technica",
                "sortBysAvailable": Array [
                  "top",
                  "latest",
                ],
                "url": "http://arstechnica.com",
                "urlsToLogos": Object {
                  "large": "",
                  "medium": "",
                  "small": "",
                },
              }
        }
    />
    <withRouter(Source)
        source={
            Object {
                "category": "general",
                "country": "us",
                "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
                "id": "associated-press",
                "language": "en",
                "name": "Associated Press",
                "sortBysAvailable": Array [
                  "top",
                ],
                "url": "https://apnews.com/",
                "urlsToLogos": Object {
                  "large": "",
                  "medium": "",
                  "small": "",
                },
              }
        }
    />
</div>,
  "nodes": Array [
    <div
      className="row"
>
      <withRouter(Source)
            source={
                  Object {
                        "category": "general",
                        "country": "au",
                        "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
                        "id": "abc-news-au",
                        "language": "en",
                        "name": "ABC News (AU)",
                        "sortBysAvailable": Array [
                          "top",
                        ],
                        "url": "http://www.abc.net.au/news",
                        "urlsToLogos": Object {
                          "large": "",
                          "medium": "",
                          "small": "",
                        },
                      }
            }
      />
      <withRouter(Source)
            source={
                  Object {
                        "category": "general",
                        "country": "us",
                        "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
                        "id": "al-jazeera-english",
                        "language": "en",
                        "name": "Al Jazeera English",
                        "sortBysAvailable": Array [
                          "top",
                          "latest",
                        ],
                        "url": "http://www.aljazeera.com",
                        "urlsToLogos": Object {
                          "large": "",
                          "medium": "",
                          "small": "",
                        },
                      }
            }
      />
      <withRouter(Source)
            source={
                  Object {
                        "category": "technology",
                        "country": "us",
                        "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
                        "id": "ars-technica",
                        "language": "en",
                        "name": "Ars Technica",
                        "sortBysAvailable": Array [
                          "top",
                          "latest",
                        ],
                        "url": "http://arstechnica.com",
                        "urlsToLogos": Object {
                          "large": "",
                          "medium": "",
                          "small": "",
                        },
                      }
            }
      />
      <withRouter(Source)
            source={
                  Object {
                        "category": "general",
                        "country": "us",
                        "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
                        "id": "associated-press",
                        "language": "en",
                        "name": "Associated Press",
                        "sortBysAvailable": Array [
                          "top",
                        ],
                        "url": "https://apnews.com/",
                        "urlsToLogos": Object {
                          "large": "",
                          "medium": "",
                          "small": "",
                        },
                      }
            }
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Sources
        sources={
                Array [
                        Object {
                          "category": "general",
                          "country": "au",
                          "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
                          "id": "abc-news-au",
                          "language": "en",
                          "name": "ABC News (AU)",
                          "sortBysAvailable": Array [
                            "top",
                          ],
                          "url": "http://www.abc.net.au/news",
                          "urlsToLogos": Object {
                            "large": "",
                            "medium": "",
                            "small": "",
                          },
                        },
                        Object {
                          "category": "general",
                          "country": "us",
                          "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
                          "id": "al-jazeera-english",
                          "language": "en",
                          "name": "Al Jazeera English",
                          "sortBysAvailable": Array [
                            "top",
                            "latest",
                          ],
                          "url": "http://www.aljazeera.com",
                          "urlsToLogos": Object {
                            "large": "",
                            "medium": "",
                            "small": "",
                          },
                        },
                        Object {
                          "category": "technology",
                          "country": "us",
                          "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
                          "id": "ars-technica",
                          "language": "en",
                          "name": "Ars Technica",
                          "sortBysAvailable": Array [
                            "top",
                            "latest",
                          ],
                          "url": "http://arstechnica.com",
                          "urlsToLogos": Object {
                            "large": "",
                            "medium": "",
                            "small": "",
                          },
                        },
                        Object {
                          "category": "general",
                          "country": "us",
                          "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
                          "id": "associated-press",
                          "language": "en",
                          "name": "Associated Press",
                          "sortBysAvailable": Array [
                            "top",
                          ],
                          "url": "https://apnews.com/",
                          "urlsToLogos": Object {
                            "large": "",
                            "medium": "",
                            "small": "",
                          },
                        },
                      ]
        }
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Sources {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "sources": Array [
            Object {
              "category": "general",
              "country": "au",
              "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
              "id": "abc-news-au",
              "language": "en",
              "name": "ABC News (AU)",
              "sortBysAvailable": Array [
                "top",
              ],
              "url": "http://www.abc.net.au/news",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
            Object {
              "category": "general",
              "country": "us",
              "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
              "id": "al-jazeera-english",
              "language": "en",
              "name": "Al Jazeera English",
              "sortBysAvailable": Array [
                "top",
                "latest",
              ],
              "url": "http://www.aljazeera.com",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
            Object {
              "category": "technology",
              "country": "us",
              "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
              "id": "ars-technica",
              "language": "en",
              "name": "Ars Technica",
              "sortBysAvailable": Array [
                "top",
                "latest",
              ],
              "url": "http://arstechnica.com",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
            Object {
              "category": "general",
              "country": "us",
              "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
              "id": "associated-press",
              "language": "en",
              "name": "Associated Press",
              "sortBysAvailable": Array [
                "top",
              ],
              "url": "https://apnews.com/",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="row"
>
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "general",
                                        "country": "au",
                                        "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
                                        "id": "abc-news-au",
                                        "language": "en",
                                        "name": "ABC News (AU)",
                                        "sortBysAvailable": Array [
                                          "top",
                                        ],
                                        "url": "http://www.abc.net.au/news",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "general",
                                        "country": "us",
                                        "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
                                        "id": "al-jazeera-english",
                                        "language": "en",
                                        "name": "Al Jazeera English",
                                        "sortBysAvailable": Array [
                                          "top",
                                          "latest",
                                        ],
                                        "url": "http://www.aljazeera.com",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "technology",
                                        "country": "us",
                                        "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
                                        "id": "ars-technica",
                                        "language": "en",
                                        "name": "Ars Technica",
                                        "sortBysAvailable": Array [
                                          "top",
                                          "latest",
                                        ],
                                        "url": "http://arstechnica.com",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "general",
                                        "country": "us",
                                        "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
                                        "id": "associated-press",
                                        "language": "en",
                                        "name": "Associated Press",
                                        "sortBysAvailable": Array [
                                          "top",
                                        ],
                                        "url": "https://apnews.com/",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
</div>,
        "_debugID": 4,
        "_renderedOutput": <div
          className="row"
>
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "general",
                                        "country": "au",
                                        "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
                                        "id": "abc-news-au",
                                        "language": "en",
                                        "name": "ABC News (AU)",
                                        "sortBysAvailable": Array [
                                          "top",
                                        ],
                                        "url": "http://www.abc.net.au/news",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "general",
                                        "country": "us",
                                        "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
                                        "id": "al-jazeera-english",
                                        "language": "en",
                                        "name": "Al Jazeera English",
                                        "sortBysAvailable": Array [
                                          "top",
                                          "latest",
                                        ],
                                        "url": "http://www.aljazeera.com",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "technology",
                                        "country": "us",
                                        "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
                                        "id": "ars-technica",
                                        "language": "en",
                                        "name": "Ars Technica",
                                        "sortBysAvailable": Array [
                                          "top",
                                          "latest",
                                        ],
                                        "url": "http://arstechnica.com",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
          <withRouter(Source)
                    source={
                              Object {
                                        "category": "general",
                                        "country": "us",
                                        "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
                                        "id": "associated-press",
                                        "language": "en",
                                        "name": "Associated Press",
                                        "sortBysAvailable": Array [
                                          "top",
                                        ],
                                        "url": "https://apnews.com/",
                                        "urlsToLogos": Object {
                                          "large": "",
                                          "medium": "",
                                          "small": "",
                                        },
                                      }
                    }
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Sources
    sources={
        Array [
            Object {
              "category": "general",
              "country": "au",
              "description": "Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
              "id": "abc-news-au",
              "language": "en",
              "name": "ABC News (AU)",
              "sortBysAvailable": Array [
                "top",
              ],
              "url": "http://www.abc.net.au/news",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
            Object {
              "category": "general",
              "country": "us",
              "description": "News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
              "id": "al-jazeera-english",
              "language": "en",
              "name": "Al Jazeera English",
              "sortBysAvailable": Array [
                "top",
                "latest",
              ],
              "url": "http://www.aljazeera.com",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
            Object {
              "category": "technology",
              "country": "us",
              "description": "The PC enthusiast's resource. Power users and the tools they love, without computing religion.",
              "id": "ars-technica",
              "language": "en",
              "name": "Ars Technica",
              "sortBysAvailable": Array [
                "top",
                "latest",
              ],
              "url": "http://arstechnica.com",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
            Object {
              "category": "general",
              "country": "us",
              "description": "The AP delivers in-depth coverage on the international, politics, lifestyle, business, and entertainment news.",
              "id": "associated-press",
              "language": "en",
              "name": "Associated Press",
              "sortBysAvailable": Array [
                "top",
              ],
              "url": "https://apnews.com/",
              "urlsToLogos": Object {
                "large": "",
                "medium": "",
                "small": "",
              },
            },
          ]
    }
/>,
}
`;
